### ISSUE TITLE
No Active Interest on JetStream Consumer Despite Connected Subscribers

### ISSUE THREAD

**Issue by tyler-eon** [nats-io/nats-server#4611]
### What version were you using?

2.10.1

### What environment was the server running in?

Alpine Linux via the `nats:2.10.1-alpine` tag.

### Is this defect reproducible?

I'm using `NACK` (the Kubernetes controller) to set up the stream and consumer so here are the YAML configurations.

Stream:

```yaml
apiVersion: jetstream.nats.io/v1beta2
kind: Stream
metadata:
  name: test-stream
spec:
  name: TEST
  subjects: ["test.*"]
  maxAge: 48h
  storage: file
  retention: interest
  discard: old
```

Consumer:

```yaml
apiVersion: jetstream.nats.io/v1beta2
kind: Consumer
metadata:
  name: test-consumer
spec:
  streamName: TEST
  durableName: test-consumer
  filterSubject: test.complete
  deliverSubject: consumer.test
  deliverGroup: test-queue-group
  deliverPolicy: all
  replayPolicy: instant
  ackPolicy: explicit
  ackWait: 30s
  maxAckPending: 1000
```

### Given the capability you are leveraging, describe your expectation?

I have the consumer set up to use a deliver group because I'm going to have multiple instances of subscribers within my Kubernetes deployment and I don't want to have duplicate messages being delivered within the cluster. I'm using the standard "core NATS" subscription flow to connect while specifying the queue group.

```
SUB consumer.test test-queue-group
```

I'm using `consumer.test` from the consumer's `deliverSubject` and `test-queue-group` from the consumer's `deliverGroup`.

I expect to be capable of subscribing with more than one client and having the messages distributed amongst the clients without duplication.

### Given the expectation, what is the defect you are observing?

When I only have clients subscribed using the specified queue group, the interest is always "none" and no messages are delivered to *any* subscribed clients.

When I create a single subscription *without* the queue group specified, suddenly there is "active interest" and messages are delivered as expected. This means that I need to have a completely separate mechanism that effectively keeps a client "dummy client" connected to the consumer subscription, minus the queue group, in order to get messages delivered.

In other words, I do:

```
SUB consumer.test test-queue-group
```

Multiple times (e.g. 5 times on different pods). No messages are delivered, but I can see multiple "unprocessed messages" according to the consumer info. Then I do:

```
SUB consumer.test
```

Just one time on a separate pod, and suddenly all of those unprocessed messages get delivered to everyone. What's also concerning is that the "dummy subscriber" I have receives messages but does not need to ACK them for the messages to be removed. As long as one of the actual subscribers using the appropriate queue group ACKs the message, it's removed from the stream as expected. It's fine behavior in this case because I don't have to worry about ACK-ing the messages immediately in the dummy subscriber, I can just ignore every message as long as it remains connected, but it's just odd behavior overall.

As for why I'm using the interest policy, it's because I do plan on having another consumer in the near future that wants to look at the same messages (`test.*`) but will need their own copy of messages and have their own retry mechanics. To my knowledge, the work queue policy won't be sufficient because *any* consumer that ACKs a message causes the message to be removed. I guess I could just create two streams with the same subject filters (if that's allowed), but I'm not certain if a work queue policy would even resolve this issue.

**Comment by tyler-eon** [comment]
I was digging into this more and I think the problem is with the `active` field in the `Consumer`. It's a channel, which I'm not familiar with exactly how those work in Go, but my guess is that it allows the field to be updated and then other systems to be notified that an update has occurred?

What the case may be, I looked at the CLI to see in which cases I get the two different messages I usually do: the no interest and the interest with group. The only thing that changes is `o.active` (`o` is a consumer struct). My guess is that calling `SUB` with a queue group name does not appropriately update this field. At least not when using the default account. I can't confirm this since I'm not proficient enough with Go to dig much deeper, but there's definitely a difference between calling `SUB` with a queue group and calling it without one.

**Comment by derekcollison** [comment]
Did you configure the consumer to be push to a DQ group?

**Comment by derekcollison** [comment]
You have to configure both of these to have it properly watch interest for the consumer.

<img width="650" alt="image" src="https://github.com/nats-io/nats-server/assets/90097/95ba6a2d-a9d2-4b36-9ae5-030fbe4aebfa">

**Comment by tyler-eon** [comment]
@derekcollison I did. If you look at the consumer configuration for NACKS:

```yaml
deliverSubject: consumer.test
deliverGroup: test-queue-group
```

I also traced the CLI call and was able to confirm that it understands this is a push subscription AND that there's a queue group associated with it.

Additionally, when subscribing from the individual clients I am passing both the subject and queue group.

```
SUB consumer.test test-queue-group
```

It successfully listens and will receive messages, and it will distribute the messages as expected across all clients with the same queue group specified, it just won't receive messages unless there is also a non-queue-group subscriber connected.

**Comment by derekcollison** [comment]
What does `nats consumer info` report?

**Comment by tyler-eon** [comment]
This was captured when I had 5 clients connected via the `SUB consumer.test test-queue-group` command:

```
Configuration:

                    Name: test-consumer
        Delivery Subject: consumer.test
          Filter Subject: test.complete
          Deliver Policy: All
     Deliver Queue Group: test-queue-group
              Ack Policy: Explicit
                Ack Wait: 30.00s
           Replay Policy: Instant
         Max Ack Pending: 1,000
            Flow Control: false

State:

  Last Delivered Message: Consumer sequence: 2,258 Stream sequence: 1,706 Last delivery: 2d12h38m51s ago
    Acknowledgment Floor: Consumer sequence: 1,815 Stream sequence: 1,815 Last Ack: 2d12h38m51s ago
        Outstanding Acks: 0 out of maximum 1,000
    Redelivered Messages: 0
    Unprocessed Messages: 2,828
         Active Interest: No interest
```

As soon as I connect a single subscriber WITHOUT the queue group specified (e.g. `SUB consumer.test`), the state changed to the following:

```
Configuration:

                    Name: test-consumer
        Delivery Subject: consumer.test
          Filter Subject: test.complete
          Deliver Policy: All
     Deliver Queue Group: test-queue-group
              Ack Policy: Explicit
                Ack Wait: 30.00s
           Replay Policy: Instant
         Max Ack Pending: 1,000
            Flow Control: false

State:

  Last Delivered Message: Consumer sequence: 5,870 Stream sequence: 16,262 Last delivery: 7.30s ago
    Acknowledgment Floor: Consumer sequence: 5,870 Stream sequence: 16,262 Last Ack: 5.69s ago
        Outstanding Acks: 0 out of maximum 1,000
    Redelivered Messages: 0
    Unprocessed Messages: 0
         Active Interest: Active using Queue Group test-queue-group
```

You can see that the stream was successfully persisting un-ACKd messages but also that none were being delivered because it *assumed* there was no one listening. As soon as I connected without a queue group specified, the 5 queue group clients were flooded with the undelivered messages and started processing and ACKing them.

It's worth noting that I manually tried to `nats sub` *with* the queue group name from the `nats-box` I've got in my cluster and that DID NOT cause messages to be delivered, i.e. it was still registered with no active interest. So I think I can say definitively that it's the presence of the queue group specifically that has some special flow which causes a lack of interest to be recorded.

Again, I'm not great with Go so I can't track down what's happening *exactly*, but my assumption is that it's got something to do with the fact I'm using a JetStream consumer. When I was subscribing directly to `test.complete` using queue groups, this was not an issue. I would get all messages as expected. But in this case I NEED to guarantee that I process each message at least once, which is why I've introduced JetStream into the flow. Prior to using JetStream this was not something that caused problems.

**Comment by Jarema** [comment]
Hey!

If you're using wire NATS protocol with queue group, to subscribe with queue group, you need to pass subject, queue-group and sid (subscription id)

```
SUB consumer.test test-queue-group 1
```

This way, it will work properly. Otherwise, what you specify as queue-group is treated as sid (subscription id) which results in  subscription without queue-group and in the result - lack of interest from the perspective of the consumer.

I suggest using any of our clients to do such tests. Go is definitely not required to work with NATS. We have dozens of clients in other languages, but the best way to test things as you do is probably by nats-cli, that we use daily ourselves.

https://github.com/nats-io/natscli

Hope that solves the issue.

**Comment by tyler-eon** [comment]
@Jarema Thanks for the response, however I should have been more explicit with the exact setup I've got. I use `gnat` (nats.ex) for the client which does indeed assign a `sid` for me. When I say `SUB x y` I really mean I'm using `Gnat.sub(:gnat, self(), subject, queue_group: queue)`.

But please note that I do mention using `nats sub` (the CLI) in my previous comment and that it continues to not work. I basically opened two terminals connected to `nats-box` (which has the CLI pre-installed) and ran `nats sub consumer.test` in one terminal and `nats consumer info`. When I provide to queue group option to `nats sub` the consumer info output still shows no active interest. When I omit the queue group option it suddenly gains interest.

TL;DR I do use a client which sends the send *and* I've tried using the CLI (`nats sub ...`) to test this behavior and it is consistent with what I'm seeing, i.e. that supplying the queue group option during subscription to a JetStream consumer does not produce active interest on its push delivery subject.

**Comment by Jarema** [comment]
Thanks for clarifications @tyler-eon.

However, I'm still not able to reproduce.
As you can see in the attached video, the consumer does not record interest if queue group is **not** specified, and does record interest when it is.

https://github.com/nats-io/nats-server/assets/7369771/0331e698-87c6-42e2-ad0a-91c03e8704cc

**Comment by tyler-eon** [comment]
Hmmm, I was able to replicate your test on a new consumer but not on my existing consumer. Even after I deleted the consumer and recreated it, it effectively stops working as soon as I try to connect actual (non-CLI) clients to it (specifically via nats.ex).

I've attached a video where you can see that the consumer I've got has no interest. When I `nats sub --queue=...`, still no interest. When I remove the `--queue` option (I mess up the first time in the video, check out the last attempt where I get the subject correct), suddenly we've got interest.

When I used `nats consumer add` I gave the test consumer the same configuration (minus the names of things) as my current consumer, so things like ACK wait times, maximum number of un-ACKd messages, etc, are all the same. Yet the test one I created via `nats consumer add` definitely worked as expected. The one created via NACK does not.

I'm curious if this is either an issue with the nats.ex client library or the NACK Kubernetes controller.

https://github.com/nats-io/nats-server/assets/1781637/4aaaef38-0327-4787-8730-65f794f801ca

**Comment by tyler-eon** [comment]
I suddenly realized that even though it's simpler for me to use push subscriptions in my distributed environment, I don't think nats.ex has any built-in distribution mechanics for the `sid`. This means that we probably have a huge overlap in clients using the same `sid`. We have like 10 NATS client subscriptions per pod and up to 10 pods at a time. The nats.ex client hard-codes the initial `sid` of new clients to one (1), so I probably have up to 100 subscriptions but reusing the same 10 `sid`s among them. So far this has *not* caused an issue, but I'm sure that I should probably seek to remedy that.

Regardless of that, I still don't know why `nats sub` (the CLI) would not switch the consumer to showing active interest.

**Comment by derekcollison** [comment]
SIDs should be scoped to connection. So ok for them to all start at 1.

**Comment by tyler-eon** [comment]
As an aside, just to check if this was a "counter issue" (i.e. a counter that was outside an acceptable range), I tried using `nats sub` in 14 different terminal windows (using the `--queue` option) and I still couldn't get my existing consumer to show active interest, needed to have at least one of those subscriptions drop the `--queue` option to register interest.

So yeah, just really weird. Not sure if there's any way to get additional information about the state of a consumer that might help debug this. As a note, it's not just like a one-time thing with this consumer configuration and application setup, I've deployed this into multiple environments and have destroyed and rebuilt these streams and consumers in our test environment and the incorrect behavior is shown consistently.

**Comment by Jarema** [comment]
What is weird here is that your consumer acts like it does not have queue group.

"normal" push consumer with a queue group will register interest only if you subscribe to it **with** queue group (as you can see in my screencast).

What is the difference in creating "my consumer" (that works as expected) and "yours" (which shows this issue behaviour)?
As even if deleting it and recreating persist the issue maybe that is a hint.

Are you creating both with NACK?

**Comment by tyler-eon** [comment]
Yes, I am creating both the Stream and the Consumer via NACK. The start of the thread has the config values posted for each.

NACK was installed via the official Helm chart: `helm install nack nats/nack -f values.yaml`.

**Comment by tyler-eon** [comment]
My hack continues to serve my purpose well enough, but I'm curious if there's any investigation being done, or planning to be done, regarding this weird bug. If not, that's alright, just wondering if I should remain hopeful or just accept the brokenness and look for alternative management of NATS (not via NACK).

**Comment by Jarema** [comment]
I'll try to reproduce it with NACK.

**Comment by Jarema** [comment]
@tyler-eon I tried to reproduce it using NACK and all works as intended.

I'm running out of ideas.
Can you share `nats stream report` and `nats consumer report`?

**Comment by tyler-eon** [comment]
Sorry for the delay, I managed to carve out some time to run the commands you suggested.

```
~ # nats stream report
Obtaining Stream stats

╭──────────────────────────────────────────────────────────────────────────────────────────╮
│                                       Stream Report                                      │
├────────┬─────────┬───────────┬───────────┬──────────┬────────┬──────┬─────────┬──────────┤
│ Stream │ Storage │ Placement │ Consumers │ Messages │ Bytes  │ Lost │ Deleted │ Replicas │
├────────┼─────────┼───────────┼───────────┼──────────┼────────┼──────┼─────────┼──────────┤
│ JOBS   │ File    │           │         1 │ 4        │ 16 KiB │ 0    │       2 │          │
╰────────┴─────────┴───────────┴───────────┴──────────┴────────┴──────┴─────────┴──────────╯

~ # nats consumer report
[default] ? Select a Stream JOBS
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                    Consumer report for JOBS with 1 consumers                                   │
├───────────────┬──────┬────────────┬──────────┬─────────────┬─────────────┬─────────────┬─────────────┬─────────┤
│ Consumer      │ Mode │ Ack Policy │ Ack Wait │ Ack Pending │ Redelivered │ Unprocessed │ Ack Floor   │ Cluster │
├───────────────┼──────┼────────────┼──────────┼─────────────┼─────────────┼─────────────┼─────────────┼─────────┤
│ jobs-consumer │ Push │ Explicit   │ 30.00s   │ 0           │ 0           │ 0           │ 980,994,086 │         │
╰───────────────┴──────┴────────────┴──────────┴─────────────┴─────────────┴─────────────┴─────────────┴─────────╯
```

Additionally, I just followed the instructions on the NACK page to set everything up. Here are the Helm values I used when installing NATS:

```yaml
config:
  jetstream:
    enabled: true

natsBox:
  enabled: true
```

And when installing NACK:

```yaml
jetstream:
  enabled: true

  nats:
    url: nats://nats:4222
```

And the Kubernetes configs are as I posted already but I'll post them again just to keep this as a mega-comment of all the configuration values.

```yaml
apiVersion: jetstream.nats.io/v1beta2
kind: Stream
metadata:
  name: jobs-stream
spec:
  name: JOBS
  subjects: ["jobs.*"]
  maxAge: 120h
  storage: file
  retention: interest
  discard: old
---
apiVersion: jetstream.nats.io/v1beta2
kind: Consumer
metadata:
  name: jobs-consumer
spec:
  streamName: JOBS
  durableName: jobs-consumer
  filterSubject: jobs.complete
  deliverSubject: group.jobs
  deliverGroup: group-jobs
  deliverPolicy: all
  replayPolicy: instant
  ackPolicy: explicit
  ackWait: 30s
  maxAckPending: 1000
```

And that's literally it. I'll also mention, in case I failed to previously, that I'm using Elixir as my application language. I'm using the official `gnat` Elixir NATS client library to connect, subscribe, and publish. I don't think the client library has been updated in a while but I also didn't see anything in the documentation to suggest that what's currently being done in the client library wouldn't work for this purpose. After all, I can still receive messages as expected, consumer groups work as expected (except in the case of push streams for some reason), and I can reply-to-ack messages. So at this point I'm also at a loss.

**Comment by Jarema** [comment]
Maybe it it's worth looking into the Elixir client for a bug related to registering interest.
Looping in @autodidaddict

**Comment by tyler-eon** [comment]
I recently upgraded all of my dependencies on their patch and minor versions, which apparently there was a recent release to 1.7 from 1.6 for the NATS Elixir client. Even though nothing in the changes between those versions implies that it would fix an issue such as this, since the upgrade I am no longer running into this issue. I'm going to take the win however I can even though it doesn't make sense.

Closing this out for now, hopefully it'll stay closed and we can just chalk this up as a weird Elixir client issue maybe?
